### ISSUE TITLE
can a fixed-length sets feature be added

### ISSUE THREAD

**Issue by lishiling5121** [scylladb/scylla#7020]
I create a table named carinfo to record car's info and there is a field named location in the table. The attribute of location field is the sets, I will write some data into the field. But some records have a lot of data in location field, the amount is more than one million, which seriously affects query performance, and worse, I only need the first 5000.
This is the carinfo table:
CREATE TABLE carinfo (
    name text PRIMARY KEY,
    location set<text>,
    ...
    PRIMARY KEY(name)
)

Is it possible to add a capped attribute for fixed length of sets, and automatically replace old data when new data is written, just like the capped attribute of mongodb?

If this attribute is added, it will greatly simplify my workflow and improve my work efficiency. I believe other users will also want to have this feature

**Comment by avikivity** [comment]
We only see the size of a set during compaction, and even then only in some cases. But it may be doable. I agree it makes some sense.

/cc @tzach

**Comment by nyh** [comment]
> We only see the size of a set during compaction, and even then only in some cases. But it may be doable. I agree it makes some sense.

Nice idea.

You mentioned "sets", but note that sets are not ordered by time, so there is no trivial way to drop specifically the **oldest** items in the set as @lishiling5121 requested. For small number of kept items (like the 5000 @lishiling5121 mentioned), a reasonable approach is to keep a heap of "newest N" items, and only output it after reading all the items of the set. This is possible but complex and inefficient.

However, as far as I can tell the MongoDB "capped collection" feature (https://docs.mongodb.com/manual/core/capped-collections/#create-a-capped-collection) is simpler and more efficient, and more similar to our **lists** (not sets), and we can implement it easily too:

1. In Scylla's lists items are normally saved time order, but have a **prepend** operation, which uses something similar to negative time so basically reversing the sort order. So we can reuse the existing list implementation to implement a capped list, and every time we want to add an item to a capped list, we add it with **prepend**.
2. Every time such a "capped list" item is read (for reading or compaction) only the first N (which are the newst items) are returned, the following ones are trivially ignored. 
3. When a coordinator reads a "capped list" from multiple replicas, they may have slightly different results that the coordinator merges - so the coordinator also needs to contain the trivial logic of dropping items from the capped set after the first N.
4. Other Scylla list operations like the (inefficient) possibility of finding the third item or deleting it, should probably not be supported for "capped list" at all. I don't think they are supported in MongoDB?

**Comment by lishiling5121** [comment]
Thank you very much for your reply. To add a capped feature on sets maybe complex and inefficient, but it's very uesful to me. In my sense, every car's `location` field should be deduplicated. Many times there is no need to duplicate data, nor too much old data.

**Comment by denesb** [comment]
> You mentioned "sets", but note that sets are not ordered by time, so there is no trivial way to drop specifically the **oldest** items in the set as @lishiling5121 requested.

Actually, in non-frozen collections, individual cells have timestamps, which we can use to order them and drop any in excess of the limit. Of course this is the mutation timestamp which is overridable by the user, so it can cause some suprises.

**Comment by nyh** [comment]
> > You mentioned "sets", but note that sets are not ordered by time, so there is no trivial way to drop specifically the **oldest** items in the set as @lishiling5121 requested.
> 
> Actually, in non-frozen collections, individual cells have timestamps, which we can use to order them and drop any in excess of the limit. Of course this is the mutation timestamp which is overridable by the user, so it can cause some suprises.

Yes, we definitely do have the *data* needed to keep only the newest items also in a set. My point was that there is no trivial way to do it as we (and MongoDB) can do it with lists. I this being possible in sets by using a heap of N kept items on every read, but that requires O(NlogN) time complexity and O(N) memory complexity on every read operation (or compaction). This may be perfectly fine for a small N (e.g., 5000).

**Comment by avikivity** [comment]
5000 is way too much for a collection.

We can require read-before-write for bounded lists (and some list operations already do a read-before-write).

**Comment by nyh** [comment]
> 5000 is way too much for a collection.

This is a good point. We have a fairly good explanation in https://docs.scylladb.com/getting-started/types/#noteworthy-characteristics on the downsides of collections. It particular, a single collection is a single row (Scylla has good support for long partitions, but not as good for long rows) and can only be retrieved as a whole. I don't know if 5000 is too much or still fine - it also depends on the length of each item, and how much you care about tail latency suffering because of these long retrievals.

An alternative to collections can be to have a partition with 5000 clustering rows, with the clustering key being reverse time (list-like) or value (set-like), and then adding a mechanism for dropping access rows in a partition during reads and compaction. It's not obvious to me how (or if) we could allow reading just a slice from such a partition - and how we can even implement _paged_ reads in the set-like case (though paging would work well in the capped list case).

> We can require read-before-write for bounded lists (and some list operations already do a read-before-write).

We can do this with LWT, but without LWT, I don't see how this can correctly implement the limit requirement in the face of concurrent updates.

**Comment by denesb** [comment]
On Wed, 2020-08-12 at 01:59 -0700, nyh wrote:
> > 5000 is way too much for a collection.
> 
> This is a good point. We have a fairly good explanation in 
> https://docs.scylladb.com/getting-started/types/#noteworthy-characteristics
> on the downsides of collections. It particular, a single collection
> is a single row (Scylla has good support for long partitions, but not
> as good for long rows) and can only be retrieved as a whole. I don't
> know if 5000 is too much or still fine - it also depends on the
> length of each item, and how much you care about tail latency
> suffering because of these long retrievals.
> 
> An alternative to collections can be to have a partition with 5000
> clustering rows, with the clustering key being reverse time (list-
> like) or value (set-like), and then adding a mechanism for dropping
> access rows in a partition during reads and compaction. It's not
> obvious to me how (or if) we could allow reading just a slice from
> such a partition - and how we can even implement paged reads in the
> set-like case (though paging would work well in the capped list
> case).

In the code we have a per-partition row limit that you can pass down in
the slice, this could be used to limit the number of clustering rows in
each partition to 5000. Not sure what the CQL syntax is.

If the clustering key was a monotonically increasing id, one could also
pair a deletion with ($recent_id - 5000) to each write to this table.

**Comment by slivne** [comment]
@tzach assigned it to you o set priority

**Comment by lishiling5121** [comment]
Thank you very much for your suggestions. After discussion, we decided to use  maps instead of sets, use month as a part of the primary key, and use the TTL to keep small amount of data. 
What is the time complexity of maps when read?

**Comment by avikivity** [comment]
O(n), where n is the number of items.
